unknown PR (#2639)
huggingface/blog: pi0.md

---
title: "π0 and π0-FAST: Vision-Language-Action Models for General Robot Control" 
thumbnail: /blog/assets/192_pi0/new_thumbnail_pi0.001.png
authors:
- user: danaaubakirova
- user: Molbap
- user: mshukor
- user: cadene

---

> We have ported the first **robot foundation models** to **Hugging Face LeRobot**! Both **π0 and π0-FAST** developed by Physical Intelligence, which are now available in the **LeRobot repository**, bringing generalist robotic intelligence to the Hugging Face ecosystem. If you are curious about how Vision-Language-Action (VLA) models differ from Vision-Language Models (VLMs) and how actions are represented, dive into this blog post to find out! 

Reviewer: ```suggestion
We have ported the first **robot foundation models** to **Hugging Face LeRobot**! Both **π0 and π0-FAST**, developed by Physical Intelligence, are now available in the **LeRobot repository**, bringing generalist robotic intelligence to the Hugging Face ecosystem. If you are curious about how Vision-Language-Action (VLA) models differ from Vision-Language Models (VLMs) and how actions are represented, dive into this blog post to find out! 
```

---
title: "π0 and π0-FAST: Vision-Language-Action Models for General Robot Control" 
thumbnail: /blog/assets/192_pi0/new_thumbnail_pi0.001.png
authors:
- user: danaaubakirova
- user: Molbap
- user: mshukor
- user: cadene

---

We have ported the first **robot foundation models** to **Hugging Face LeRobot**! Both **π0 and π0-FAST** developed by Physical Intelligence, which are now available in the **LeRobot repository**, bringing generalist robotic intelligence to the Hugging Face ecosystem. If you are curious about how Vision-Language-Action (VLA) models differ from Vision-Language Models (VLMs) and how actions are represented, dive into this blog post to find out! 

> [Huggingface collection of Pi0 models](https://huggingface.co/collections/lerobot/pi0-models-67a0f92dc62e52ace7220eba) | [LeRobot] (link)

Reviewer: Still not up, is it?

Reviewer 2: not yet! 30min

Developing generalist robot policies, or robot foundation models, presents three key challenges:

1. **The need for large-scale research** to fully leverage pre-training benefits.
2. **Designing model architectures** that can integrate diverse data sources while capturing complex physical interactions. A key challenge in this regard is **cross-embodiment training**, where a model must learn from diverse robot types with varying configurations, control spaces, and action representations. Existing approaches tackle this by:
   - **Combining multimodal datasets** from different robotic platforms to enhance generalization.
   - **Using shared representations** to bridge the gap between distinct robot morphologies, such as single-arm, dual-arm, and mobile manipulators.

3. **Crafting an effective training recipe**, as recent advances in NLP and vision have heavily relied on careful pre-training and post-training strategies. 

In this post, we introduce **π0 and π0-FAST**, prototype models and learning frameworks developed by **Physical Intelligence**, designed to overcome these challenges.

---

> ## 🔍 What is π0?

Reviewer: I think there are many bold terms in this section. Personally I find it somewhat distracting, but it's totally your call.

[Paper](https://www.physicalintelligence.company/download/pi0.pdf) | [Jax Code](https://github.com/Physical-Intelligence/openpi)

π0 (**Pi-Zero**) is a **Vision-Language-Action (VLA) model**, developed by the [Physical Intelligence team](https://www.physicalintelligence.company) designed for **generalist robot control**. It builds upon **large-scale pretraining** and **flow matching-based action generation**, enabling robots to perform **dexterous manipulation tasks** across different embodiments.

π0 is trained on data from **7 robotic platforms** and **68 unique tasks**, demonstrating strong **zero-shot** and **fine-tuned performance** on complex, real-world tasks such as **laundry folding, table bussing, grocery bagging, box assembly, and object retrieval**.

> Unlike standard robotic policies, **π0 employs flow matching** to produce **smooth, real-time action trajectories at 50Hz**, making it highly **efficient, precise, and adaptable** for real-world deployment.

Reviewer: We don't need to explain how flow matching works, but perhaps it'd be helpful to contextualize a bit as one recent technique behind some great quality improvements in diffusion models.

## How to Use π0 in LeRobot?

> 

Reviewer 2: In line with @pcuenca I'd add something like this (needs the snippet @Cadene )
```suggestion
π0 models are foundational models that, much like PaliGemma, are made to be adapted to a variety of frameworks, environments, and scene inputs. The base models here are usable as-is, in particular π0. 

### Inference on π0 pretrained model
#### add python snippet...
However, the performances are reduced as it's a conversion from jax to torch and from a specific environment. We recommend fine-tuning your own π0 to your own environment, like below. 
```

Reviewer 2: suggestions are broken with code snippets :skull: but seems fine

> Unlike standard robotic policies, **π0 employs flow matching** to produce **smooth, real-time action trajectories at 50Hz**, making it highly **efficient, precise, and adaptable** for real-world deployment.

Reviewer 2: for flow matching @pcuenca @danaaubakirova 
```suggestion
Unlike standard robotic policies, **π0 employs flow matching** to produce **smooth, real-time action trajectories at 50Hz**, making it highly **efficient, precise, and adaptable** for real-world deployment. Flow matching was used in continuous normalizing flows and improved generation quality in diffusion models. The denoising process π0 used works in the same way, starting with a random noise that progressively converges towards a sequence of motor actions that make sense. 
```

> ### Fine-tuning the π0 Pretrained Model

Reviewer: Do we always need to fine-tune before use? Otherwise, we could explain how to use a fine-tuned model, and then the benefits of fine-tuning.

> To fine-tune the **π0** model using the `pi0_base` checkpoint from `openpi`, run the following command:

Reviewer: Links here would be nice :)

```python
python lerobot/scripts/train.py \
--policy.path=lerobot/pi0 \
--dataset.repo_id=danaaubakirova/koch_test
```

> To fine-tune the π0 neural network with PaliGemma and Expert Gemma, which were pretrained using VLM default parameters before π0 fine-tuning, execute:

Reviewer 2: explained above in suggestion!

  - **State tokens** (both current and previous),
  - **Other action tokens**.

#### **Prefix Tokens**
- Represent the **full scene** and fully attend to each other, similar to **PaliGemma**.

### **Key Idea**
These tokens encapsulate:
- The **robot’s internal representation** of the environment (**state**),
- The **commands or controls** the robot issues (**action**),
- An encoding of **time or step index** (**time embedding**).

They are **appended after** the prefix portion (images + text), so the **prefix serves as context** (e.g., a scene image, language instructions like *"be a good robot"* or *"transfer the cube"*), while the **suffix captures policy‐specific features**.

---

## ⚡ Towards Faster attention in π0 

The particular shape of the attention mask in pi0 brings some interesting challenges in how attention is computed. Let's jump into the details!

### **Handling 2D Attention Masks**
The resulting **2D causal mask** exhibits strong **block sparsity**, but defining the boundaries of each block—especially in a batch of samples—is a bit tricky. We are used to causal masks with triangular structures for autoregressive modeling, but this is not one of these cases. 

> As you can see in this example below: the image (first element) has some padding tokens, representing empty cameras. Then, text tokens are added, with text tokens as well. This "prefix" part forms a fully noncausal attention, as in PaliGemma. Then, the suffix (state + action/time tokens) has a causal-block structure. The eager naive implementation matmuls and softmaxes over all of this, which is quite inefficient.

Reviewer: ```suggestion
As you can see in this example below: the image (first element) has some padding tokens, representing empty cameras. Then, text tokens are added, with state tokens as well. This "prefix" part forms a fully noncausal attention, as in PaliGemma. Then, the suffix (state + action/time tokens) has a causal-block structure. The eager naive implementation matmuls and softmaxes over all of this, which is quite inefficient.
```

I guess? Or should we remove?

<div align="center">
    <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/QXPQXYQFQbM_zada-VSw0.png" alt="VLA Attention Mask" style="width: 55%; border: none;">
</div>

<p align="center">

>   <em> Figure 2: The visualization of the VLA attention mask </em>

Reviewer: ```suggestion
  <em> Figure 1: The visualization of the VLA attention mask </em>
```

```python
python lerobot/scripts/train.py \
--policy.path=lerobot/pi0 \
--dataset.repo_id=danaaubakirova/koch_test
```

To fine-tune the π0 neural network with PaliGemma and Expert Gemma, which were pretrained using VLM default parameters before π0 fine-tuning, execute:

```python
python lerobot/scripts/train.py \
--policy.type=pi0 \
--dataset.repo_id=danaaubakirova/koch_test
```

You can also use the pretrained π0 model independently from the LeRobot training framework with the following code:

```python
policy = Pi0Policy.from_pretrained("lerobot/pi0")
```

## What is the difference between VLMs and VLAs?

Vision-Language Models (VLMs) and Vision-Language-Action Models (VLAs) share a common foundation: transformers. However, the key distinction lies in action representation. While VLMs process and generate multimodal representations (images and text), VLAs extend this by incorporating action and observation state tokens. With these additional tokens in place, the next challenge is understanding how attention is computed.

## Attention Mechanisms in Robotics Policies

Let’s expand our vocabulary and introduce key terms:

> #### **State Tokens**

Reviewer 2: ```suggestion
#### **State Token**
```

> - Represent the robot’s **current environment state** (e.g., joint angles, sensor values, or other relevant observations).

Reviewer 2: ```suggestion
- It is a single token that represents the robot’s **current environment state** (e.g., joint angles, sensor values, or other relevant observations).
```

mask_mod = precomputed_mask_factory(causal_mask)
# create a block mask with that signature
block_mask = create_block_mask(
	mask_mod=mask_mod,
	# ...
)

# Call flex attention now!
attn_output, attention_weights = flex_attention(
	query,
	key,
	value,
	block_mask=block_mask,
)

```

The current implementation runs, and a WIP is to have it support `torch.compile` and leverage it to the fullest!

##  How to effectively represent Actions?

Now that we know actions are simply n-dimensional vectors that can be tokenized, we can explore the challenges of action representation in Vision-Language-Action (VLA) models. The way actions are represented directly impacts **efficiency, generalization, and execution fidelity**.

One approach is **semantic action representation**, where actions are described as **high-level concepts** like sub-tasks or keypoints. While this allows for few-shot and zero-shot learning, it often relies on hand-designed low-level controllers, limiting flexibility across different robots. In contrast, low-level control representations map actions directly to motor commands, enabling precise movements but making training **less stable and harder to scale**.

> Most existing VLAs use **discrete action tokenization**, converting continuous actions into discrete tokens generated autoregressively. The most common method—per-dimension, per-timestep binning—struggles with high-frequency control tasks, leading to lossy representations and **inefficient training**. Alternatives like vector quantization (VQ) and time-series compression help, but **VQ is sensitive to hyperparameters**, making it less reliable for diverse robot designs.

Reviewer 2: ```suggestion
Most existing VLAs use **discrete action tokenization**, converting continuous actions into discrete tokens generated autoregressively. The most common method—meaning per-dimension and per-timestep binning—struggles with high-frequency control tasks, leading to lossy representations and **inefficient training**. Alternatives like vector quantization (VQ) and time-series compression help, but **VQ is sensitive to hyperparameters**, making it less reliable for diverse robot designs.
```

To address these issues, **Frequency-space Action Sequence Tokenization (FAST)** introduces a **novel time-series compression approach** using the **Discrete Cosine Transform (DCT)**. FAST **reduces redundancy**, **improves efficiency**, and **enhances action fidelity**. 

With this, we present **π0-FAST**, faster and autoregressive version of **π0** also available in Lerobot repo, an extension of **π0**, which leverages this **new tokenizer** for **better action representation**.

---
## 🚀 What is π0-FAST?
[Paper](https://arxiv.org/pdf/2501.09747) | [Jax Code](https://github.com/Physical-Intelligence/openpi)

π0-FAST is an **autoregressive version** of π0, introducing **FAST (Frequency-space Action Sequence Tokenization)**—a new tokenization scheme that enhances efficiency and performance.

### Key Advantages of π0-FAST:
- **5x faster training** compared to diffusion-based VLAs.
- **Improved action representation**, reducing redundancy in action sequences.
- **Stronger generalization** across unseen environments and robot morphologies.

🔗 The **π0-FAST tokenizer** can be accessed here: [FAST Tokenizer](https://huggingface.co/physical-intelligence/fast)

---

> FAST uses the Discrete Cosine Transform (DCT) to compress continuous action sequences into discrete tokens. The process, illustrated in Figure 2, begins with **normalizing raw robot actions**, mapping the **1st and 99th quantiles** of each action dimension to the range **[-1,1]**. This normalization ensures consistency across different robotic systems and improves robustness against outliers. 
> 
> Each action dimension is then **transformed independently** using DCT, converting the time-domain signal into the frequency domain. To reduce redundancy, **insignificant coefficients** are removed through a **scale-and-round operation**, where a hyperparameter balances **compression rate and reconstruction accuracy**. The resulting **DCT coefficient matrix**, often sparse, is flattened into a **one-dimensional sequence of integers**, interleaving **low-frequency components first** across dimensions to preserve critical information.
> 
> To further compress the sequence, **Byte Pair Encoding (BPE)** is applied. BPE merges frequently occurring patterns across dimensions while maintaining a **fixed-size vocabulary**, ensuring seamless integration with **vision-language models** used in robotics.
> 
> ![image/png](https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/w3p752hK9lyXH1HHYQfde.png)
> <p align="center">
>   <em> Figure 2: The FAST action tokenization pipeline </em>
> </p> 
> 
> Since all operations are **invertible**, actions can be **reconstructed efficiently and losslessly** from tokens. The tokenization pipeline has only **two hyperparameters**: the **scaling coefficient** applied before rounding and the **BPE vocabulary size**. Both parameters remain **robust across different datasets**.
> 
> Additionally, a **universal version of FAST**, called **FAST+**, has been trained on **one million action sequences** from **single-arm, bimanual, and mobile manipulation robots**, making it applicable across diverse robotic setups. FAST+ is available as a **Hugging Face AutoProcessor**, allowing users to tokenize action sequences with just **a few lines of code**. 
> 
> For **optimal compression**, input actions should be **quantile-normalized to [-1,1]** before tokenization. The **AutoProcessor module** also enables users to **train a custom FAST tokenizer** on their own datasets with minimal effort.

Reviewer 2: ```suggestion
**FAST** uses the Discrete Cosine Transform (DCT) to compress continuous action sequences into discrete tokens. The process, illustrated in Figure 2, begins with normalizing raw robot actions, mapping the 1st and 99th quantiles of each action dimension to the range [-1,1]. This normalization ensures consistency across different robotic systems and improves robustness against outliers. 

Each action dimension is then transformed independently using DCT, converting the time-domain signal into the frequency domain. To reduce redundancy, insignificant coefficients are removed through a scale-and-round operation, where a hyperparameter balances compression rate and reconstruction accuracy. The resulting DCT coefficient matrix, often sparse, is flattened into a one-dimensional sequence of integers, interleaving low-frequency components first across dimensions to preserve critical information.

To further compress the sequence, Byte Pair Encoding (BPE) is applied. As usual, BPE merges frequently occurring patterns across dimensions while maintaining a fixed-size vocabulary.

![image/png](https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/w3p752hK9lyXH1HHYQfde.png)
<p align="center">
  <em> Figure 2: The FAST action tokenization pipeline </em>
</p> 

Since all operations are invertible, actions can be reconstructed efficiently and losslessly from tokens. The tokenization pipeline has only two hyperparameters: the scaling coefficient applied before rounding and the BPE vocabulary size. Both parameters remain robust across different datasets.

Additionally, a universal version of FAST, called **FAST+**, has been trained on one million action sequences from single-arm, bimanual, and mobile manipulation robots, making it applicable across diverse robotic setups. FAST+ is available as a **Hugging Face AutoProcessor**, allowing users to tokenize action sequences with just a few lines of code. 

For optimal compression, input actions should be quantile-normalized to [-1,1] before tokenization. The AutoProcessor module also enables users to train a custom FAST tokenizer on their own datasets with minimal effort.
```
